### PR TITLE
Reduced checking for new app changes based on user's idleness

### PR DESCRIPTION
**Pull request by orangejenny**
Should still check approximately every 20 sec while the user is active.

@dannyroberts / @mkangia 

`w=1`

### CODE REVIEW

**Review by millerdev — APPROVED** [review]

File: corehq/apps/app_manager/static/app_manager/js/menu.js [hunk 1]
```diff
@@ -32,19 +32,39 @@ hqDefine("app_manager/js/menu", [
...
+            },
+            updateLastActivity = function () {
+                if (msSinceLastActivity() > frequency) {
+                    // If they're coming back after long inactivity, do an immediate check
+                    _checkPublishStatus(true);
+                }
+                lastActivity = (new Date()).getTime();
+            };
+
+        updateLastActivity();
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/menu.js]: When I first wrote this, the expression `(new Date()).getTime()` was splashed all over and it was hard to read, so I removed all but one of them. Happy to put one back to initialize `lastActivity` if that's easier to understand than mentally running through the undefined/NaN logic.

File: corehq/apps/app_manager/static/app_manager/js/menu.js [hunk 2]
```diff
@@ -32,19 +32,39 @@ hqDefine("app_manager/js/menu", [
         });
     };
 
+    // Frequently poll for changes to app, for the sake of showing the "Updates available to puiblish" banner.
```
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/menu.js]: nit: `puiblish`

File: corehq/apps/app_manager/static/app_manager/js/menu.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/menu.js]: 3fbb14e4e0e

File: corehq/apps/app_manager/static/app_manager/js/menu.js (hunk 2 above)
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/menu.js]: 278631bb61f

File: corehq/apps/app_manager/static/app_manager/js/menu.js [hunk 3]
```diff
@@ -32,19 +32,39 @@ hqDefine("app_manager/js/menu", [
...
+            updateLastActivity = function () {
+                if (msSinceLastActivity() > frequency) {
+                    // If they're coming back after long inactivity, do an immediate check
+                    _checkPublishStatus(true);
+                }
+                lastActivity = (new Date()).getTime();
+            };
+
+        updateLastActivity();
+        $(document).on('mousemove keydown', _.throttle(updateLastActivity, 5000));
```
Comment (mkangia) [inline corehq/apps/app_manager/static/app_manager/js/menu.js]: `mousemove` would make this get called a lot number of times, not sure how that effects JS performance.
Can we instead use a way to do this when the tab is active, https://stackoverflow.com/questions/1760250/how-to-tell-if-browser-tab-is-active?

**Comment by mkangia** [comment]
Overall,
1. Would need to dive in, but if you can tell what is the use of this check eventually?
2. Can we reduce the frequency, would depend on answer to the above point?

**Review by millerdev — APPROVED** [review]

File: corehq/apps/app_manager/static/app_manager/js/menu.js (hunk 3 above)
Comment (orangejenny) [reply corehq/apps/app_manager/static/app_manager/js/menu.js]: Had trouble getting it working earlier. Updated: f7039a808b8

**Comment by orangejenny** [comment]
@mkangia This check is what causes this banner to pop up when changes are made in the app:
screen shot 2019-02-08 at 1 57 32 pm

This was added in app manager v2 because we've gotten feedback about users wanting to know when someone else is working in the app. The frequency is high to make it feel real/immediate.

**Comment by orangejenny** [comment]
I'm not wholly opposed to reducing the frequency, but it'll make the feedback feel less immediate, so I'd rather start by reducing unnecessary checks.

**Review by mkangia — APPROVED** [review]

**Comment by mkangia** [comment]
Thanks @orangejenny 
i was wondering the benefit to the user compared to the cost we take to produce that information.

`latestReleasedBuild` is not used in the JS, we can skip fetching that for this AJAX at least. i see it's needed for `get_released_app_version` so we can't remove it entirely.

Probably out of scope of this PR,
if thats what we need to tell the user,
if we could just
1. on app.save add a property called 'changes_available' to true on app
2. on save_copy method when me make a build, put that as false
or
1. compare last_modified with a timestamp last_build(not present currently)

and just check on that instead of finding
1. current app version
2. last build
3. last released build(unused anyway)

**Comment by orangejenny** [comment]
@mkangia Yeah, it's definitely a tradeoff - likely worthwhile if you're on a team actively app building and couch is doing okay, not worth it if you're the only person working on your app and/or couch is performing badly. Another possibility would be to do the check against ES, which already has the `build_spec` properties and `built_on` available (although the delay in data reaching ES might be an issue). I'm going to merge this one, and I imagine we'll come back to this logic in the future.

**Comment by mkangia** [comment]
thanks @orangejenny 

```
`latestReleasedBuild` is not used in the JS, we can skip fetching that for this AJAX at least. i see it's needed for `get_released_app_version` so we can't remove it entirely.
```
Just checking that you chose not to make this change? Totally fine if you did, just confirming that it was not missed.

**Comment by orangejenny** [comment]
@mkangia Yep, that'd be a good improvement to make but I'm calling it out of scope for this PR.

**Pull request merged**